diff --git a/includes/entityconnect.form.inc b/includes/entityconnect.form.inc
index 81c29e5..9ac4c4c 100644
--- a/includes/entityconnect.form.inc
+++ b/includes/entityconnect.form.inc
@@ -99,17 +99,33 @@ function _entityconnect_field_attach_form($entity_type, $entity, &$form, &$form_
   // because "Edit" button will not react well on multiple selected items.
   // Autocomplete widget has not #type value, so we are testing it
   // via $widget_container['#type'].
-  // We also test if $widget_container['view'] is set to support
-  // the widget provided by entityreference_view_widget module.
-  if (isset($widget_container['#type']) || isset($widget_container['view'])) {
+  // We also test if $instance['widget']['type'] is
+  // 'entityreference_view_widget' to support entityreference_view_widget
+  // module.
+  if ($instance['widget']['type'] == 'entityreference_view_widget') {
+    isset($parents) ? $parents : $parents = '';
+    $extra_class = (!isset($widget_container['#cardinality']) || $widget_container['#cardinality'] == 1) ? ' single-value' : ' multiple-values';
+    if (user_access('entityconnect add button') && $addbuttonallowed == 0) {
+      entityconnect_add_form_element_new($field_container, $language, $field_name, $parents, 'all', $extra_class, $instance, $entity_type, $entity, $acceptable_types);
+    }
+    if (user_access('entityconnect edit button') && $editbuttonallowed == 0) {
+      foreach ($widget_container as $key => &$widget_element) {
+        if (is_numeric($key)) {
+          isset($parents) ? $parents : $parents = '';
+          entityconnect_add_form_element_edit($widget_element, $language, $field_name, $parents, $key, $extra_class, $entity_type, $entity);
+        }
+      }
+    }
+  }
+  elseif (isset($widget_container['#type'])) {
     isset($parents) ? $parents : $parents = '';
     $extra_class = isset($widget_container['#type']) ? $widget_container['#type'] : 'autocomplete';
     $extra_class .= (!isset($widget_container['#cardinality']) || $widget_container['#cardinality'] == 1) ? ' single-value' : ' multiple-values';
     if (user_access('entityconnect add button') && $addbuttonallowed == 0) {
-      entityconnect_add_form_element_new($field_container, $language, $field_name, $parents, 'all', $extra_class, $instance, $entity_type, $acceptable_types);
+      entityconnect_add_form_element_new($field_container, $language, $field_name, $parents, 'all', $extra_class, $instance, $entity_type, $entity, $acceptable_types);
     }
     if (user_access('entityconnect edit button') && $editbuttonallowed == 0) {
-      entityconnect_add_form_element_edit($field_container, $language, $field_name, $parents, 'all', $extra_class, $entity_type);
+      entityconnect_add_form_element_edit($field_container, $language, $field_name, $parents, 'all', $extra_class, $entity_type, $entity);
     }
   }
   else {
@@ -122,11 +138,11 @@ function _entityconnect_field_attach_form($entity_type, $entity, &$form, &$form_
       if (is_numeric($key)) {
         if (user_access('entityconnect add button') && $addbuttonallowed == 0) {
           isset($parents) ? $parents : $parents = '';
-          entityconnect_add_form_element_new($widget_element, $language, $field_name, $parents, $key, $extra_class, $instance, $entity_type, $acceptable_types);
+          entityconnect_add_form_element_new($widget_element, $language, $field_name, $parents, $key, $extra_class, $instance, $entity_type, $entity, $acceptable_types);
         }
         if (user_access('entityconnect edit button') && $editbuttonallowed == 0) {
           isset($parents) ? $parents : $parents = '';
-          entityconnect_add_form_element_edit($widget_element, $language, $field_name, $parents, $key, $extra_class, $entity_type);
+          entityconnect_add_form_element_edit($widget_element, $language, $field_name, $parents, $key, $extra_class, $entity_type, $entity);
         }
       }
     }
@@ -136,7 +152,7 @@ function _entityconnect_field_attach_form($entity_type, $entity, &$form, &$form_
 /**
  * Here we attach a "Add" submit button.
  */
-function entityconnect_add_form_element_new(&$widget_element, $language, $field_name, $parents, $key, $extra_class, $instance, $entity_type, $acceptable_types = NULL) {
+function entityconnect_add_form_element_new(&$widget_element, $language, $field_name, $parents, $key, $extra_class, $instance, $entity_type, $entity, $acceptable_types = NULL) {
   $widget_element["add_entityconnect__{$field_name}_{$key}_{$parents}"] = array(
     '#type' => 'submit',
     '#limit_validation_errors' => array(),
@@ -146,6 +162,7 @@ function entityconnect_add_form_element_new(&$widget_element, $language, $field_
     '#suffix' => '</div>',
     '#key' => $key,
     '#field' => $field_name,
+    '#entity' => $entity,
     '#entity_type_target' => $entity_type,
     '#acceptable_types' => $acceptable_types,
     '#add_child' => TRUE,
@@ -166,7 +183,7 @@ function entityconnect_add_form_element_new(&$widget_element, $language, $field_
 /**
  * Here we attach a "Edit" submit button.
  */
-function entityconnect_add_form_element_edit(&$widget_element, $language, $field_name, $parents, $key, $extra_class, $entity_type) {
+function entityconnect_add_form_element_edit(&$widget_element, $language, $field_name, $parents, $key, $extra_class, $entity_type, $entity) {
   $widget_element["edit_entityconnect__{$field_name}_{$key}_{$parents}"] = array(
     '#type' => 'submit',
     '#limit_validation_errors' => array(array($field_name)),
@@ -176,6 +193,7 @@ function entityconnect_add_form_element_edit(&$widget_element, $language, $field
     '#suffix' => '</div>',
     '#key' => $key,
     '#field' => $field_name,
+    '#entity' => $entity,
     '#entity_type_target' => $entity_type,
     '#add_child' => FALSE,
     '#language' => $language,
@@ -226,7 +244,7 @@ function entityconnect_add_edit_button_submit(&$form, &$form_state) {
             }
           }
           else {
-            $target_id = $value;
+            $target_id = array_shift($target_id);
           }
         }
       }
@@ -234,7 +252,7 @@ function entityconnect_add_edit_button_submit(&$form, &$form_state) {
 
     // TODO : Find the original problem of the entity connect use with
     // Field Collection.
-    if (!empty($target_id)) {
+    if (!is_array($target_id) && !empty($target_id)) {
       // Take "label (entity id)', match the id from parenthesis.
       if (preg_match("/.+\((\d+)\)/", $target_id, $matches)) {
         $target_id = $matches[1];
@@ -391,6 +409,9 @@ function entityconnect_return_form_alter(&$form, &$form_state, $form_id, $cache_
     $form_state['rebuild'] = TRUE;
     $data = $cache->data;
     $parents = isset($form_state['clicked_button']['#parents']) ? $form_state['clicked_button']['#parents'] : NULL;
+    $entity = $form_state['clicked_button']['#entity'];
+    $entity_type_target = $form_state['clicked_button']['#entity_type_target'];
+    $field_name = $form_state['clicked_button']['#field'];
 
     unset($form_state['clicked_button']);
     unset($form_state['submit_handlers']);
@@ -401,9 +422,15 @@ function entityconnect_return_form_alter(&$form, &$form_state, $form_id, $cache_
       $language = isset($field_container['#language']) ? $field_container['#language'] : LANGUAGE_NONE;
     }
 
-    isset($language) ? $widget_container = &$field_container[$language] : $widget_container = $field_container;
-    isset($widget_container['#type']) ? $widget_container_type = $widget_container['#type'] : $widget_container_type = 'autocomplete';
+    isset($field_container[$language]) ? $widget_container = &$field_container[$language] : $widget_container = $field_container;
 
+    // Get widget type from field instance.
+    if (!empty($entity_type_target) && !empty($entity)) {
+      list(, , $bundle) = entity_extract_ids($entity_type_target, $entity);
+      $instance = field_info_instance($form['#entity_type'], $field_name, $bundle);
+    }
+    isset($instance['widget']['type']) ? $widget_container_type = $instance['widget']['type'] : $widget_container_type = 'autocomplete';
+ 
     $field_info = field_info_field($data['field']);
 
     if (isset($data['target_id']) && $form_state['rebuild'] === TRUE) {
@@ -444,6 +471,11 @@ function entityconnect_return_form_alter(&$form, &$form_state, $form_id, $cache_
                 $element['#default_value'] = $widget_container['#value'] + array($element['#default_value'] => $element['#default_value']);
                 break;
 
+                // Entity reference views widget.
+              case 'entityreference_view_widget':
+                $element['#default_value'] = $node->nid;
+                break;
+
               default:
                 if ($field_info['module'] == 'entityreference') {
                   $element['#default_value'] = $node->title;
@@ -707,6 +739,9 @@ function _entityconnect_alter_form_state_input(&$form_state, $module, $widget_ty
     case 'radios':
     case 'checkboxes':
       array_push($parents, $language);
+    case 'entityreference_view_widget':
+      array_push($parents, NULL);
+      array_push($parents, "target_id");
       break;
 
     default:
